File: core/PalindromeLinkedList/PalindromeLinkedList.java
package com.dbc;

public class PalindromeLinkedList {
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public boolean isPalindrome(ListNode head) {
        ListNode slow = head, fast = head;
        while (slow.next != null && fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode prev = null, cur = slow.next;
        while (cur != null) {
            ListNode temp = cur.next;
            cur.next = prev;
            prev = cur;
            cur = temp;
        }
        while (head != null && prev != null) {
            if (head.val != prev.val) return false;
            head = head.next;
            prev = prev.next;
        }
        return true;
    }
}

File: core/PalindromeLinkedList/PalindromeLinkedList.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def isPalindrome(head: ListNode) -> bool:
    count, cur = 0, head
    while cur:
        count += 1
        cur = cur.next
    if count == 1:
        return True

    prev, cur = None, head
    for i in range(count // 2):
        temp = cur.next
        cur.next = prev
        prev = cur
        cur = temp

    if count % 2 == 1:
        cur = cur.next
    while prev and cur:
        if prev.val != cur.val:
            return False
        prev = prev.next
        cur = cur.next

    return True




